docs/RFCS/20191009_gc_protected_timestamps.md

```diff
@@ -0,0 +1,597 @@
+timestamps. To aid in this process we allow clients to associate metadata with
+each protected timestamp.
+
+The decision allows different users of protected timestamps to ensure that they
+are ultimately cleaned up in a manor which is appropriate for their use case. If
+a client has confidence that the lifecycle of some external state will move to
+a terminal state and can always release a protected timestamp in that
+transition, great. If the client would prefer to use metadata to associate 
+protected timestamps to some other state and run periodic reconciliation,
+also great.
+
+# Reference-level explanation
+
+In this initial implementation we'll say that the main creators of protected
+timestamps are schema changes and backups. In both cases these clients these
+clients generally have some decided upon timestamp and need to ensure that
+that timestamp stays alive throughout the duration of the job, even if the
+job is paused. For each of these cases there is a one-to-one mapping from a job
+to a protected timestamp.
+
+This section will proceed by introducing the interface with
+which job implementers create and release protected timestamps. Then we'll
+introduce the interface exposed to the `storage` layer for use
+during the GC process. Lastly we'll walk through the interface exposed for
+creating mechansism to periodically reconcile protected timestamps with
+external state. After the interfaces have been exposed, the section will dive
+into a more detailed design of their implementation.
+
+## Interfaces
+
+To avoid circular dependencies on SQL, the implementation of the interfaces
+exported by `protectedts` will be implemented by a sub-package. Furthermore
+another package `protectedtsbase` which `storage` can import will contain
+basic structs and interfaces for its consumption.
+
+### `protectedts.Store`
+
+Jobs will interact with a `protectedts.Store`:
+
+```go
+package protectedts 
+
+// Store provides clients with a mechanism to protect and release
+// protected timestamps for a set of spans.
+//
+// Clients may provide a Txn object which will allow them to write the id
+// of this new protection transactionally into their own state. If the caller
+// provides a Txn object then it is the caller's responsibility to call
+// CleanupOnError if an error is returned. If an error is returned and the
+// transaction is subsequently committed, the resulting state of the system
+// is undefined (perhaps this is a good place for savepoints?).
+//
+// It is the caller's responsibility to ensure that a timestamp is ultimately
+// released.
+type Store interface {
+
+  // Protect will durably create a protected timestamp, if no error is returned
+  // then no data in the specified spans at which are live at the specified
+  // timestamp can be garbage collected until the this ProtectedTimestamp is
+  // released.
+  //
+  // Protect will set the deadline on the provided transaction to ensure that
```

Comment:
Setting a deadline on an existing txn object seems not ideal. Are you doing this only to avoid touching the ranges? I think if the zone config GC TTL was recently increased you'll still get burned, so you need to touch all ranges anyways (to verify that they see the ... latest? zone config, and have a GC TTL that is still compatible), so I hope you can drop the deadline instead.